src/irmin/tree.ml

```diff
@@ -1841,69 +1848,93 @@ module Make (P : Private.S) = struct
           let v = P.Node.Val.of_proof p in
           P.Node.Key.hash v
 
-    let rec to_tree (p : t) : tree =
+    let of_tree t = of_tree t Fun.id
+
+    let rec to_tree : type a. t -> (tree -> a) -> a =
+     fun p k ->
       match p with
-      | Blinded_node h -> `Node (Node.of_hash None h)
-      | Blinded_contents (c, h) -> `Contents (Contents.of_hash None c, h)
-      | Node n -> tree_of_proof_steps n
-      | Inode { length; proofs } -> tree_of_inode length proofs
-
-    and tree_of_proof_steps n : tree =
-      let bindings =
-        List.to_seq n
-        |> Seq.map (fun (s, p) ->
-               let n = to_tree p in
-               (s, n))
-        |> StepMap.of_seq
+      | Blinded_node h -> k (`Node (Node.of_hash None h))
+      | Blinded_contents (c, h) -> k (`Contents (Contents.of_hash None c, h))
+      | Node n -> tree_of_node n k
+      | Inode { length; proofs } -> tree_of_inode length proofs k
+
+    and tree_of_node : type a. (step * t) list -> (tree -> a) -> a =
+     fun n k ->
+      let rec aux acc = function
+        | [] -> k (`Node (Node.of_map acc))
+        | (s, p) :: rest -> to_tree p (fun n -> aux (StepMap.add s n acc) rest)
       in
-      `Node (Node.of_map bindings)
+      aux StepMap.empty n
 
     (** [tree_of_inode] is solely called on the root of an inode tree *)
-    and tree_of_inode len proofs : tree =
-      let elts =
+    and tree_of_inode : type a. int -> (int * t) list -> (tree -> a) -> a =
```

Comment:
This level of recursion is not in CPS, but I guess this is alright

Reply:
Fixed in e9d9b0b

Reply:
That's better. However, the `tree_of_inode` frame still remains during de computation of the subtrees. The the last 2 lines of `tree_of_inode` should be in the CPS callback:
```
      List.iter (fun (s, elt) -> Node.add_to_findv_cache n s elt) elts;
      k (`Node n)
```

Reply:
thanks, should be fixed in e9d9b0b

Reply:
LGTM